src/main/java/com/auth0/jwk/JwkSet.java

```diff
@@ -0,0 +1,78 @@
+package com.auth0.jwk;
+
+import java.util.Collections;
+import java.util.List;
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Represents a set of JSON Web Keys (JWK) used to verify the signature of JWTs.
+ */
+public class JwkSet {
+
+    private final List<Jwk> keys;
+
+    /**
+     * Create a new empty JWK set.
+     */
+    public JwkSet() {
+        this(Collections.<Jwk>emptyList());
+    }
+
+    /**
+     * Creates a new JWK set with a single key.
+     * @param key the JWK. Must not be {@code null}.
+     */
+    public JwkSet(final Jwk key) {
+        this(Collections.singletonList(key));
+        checkArgument(key != null, "The JWK must not be null");
+    }
+
+    /**
+     * Creates a new JWK set with a list of keys.
+     * @param keys the list of keys. Must not be {@code null}.
+     */
+    public JwkSet(final List<Jwk> keys) {
+        checkArgument(keys != null, "The JWK list must not be null");
+        this.keys = Collections.unmodifiableList(keys);
+    }
+
+    /**
+     * Get the keys of this JWK set.
+     * @return The keys, or an empty list if there are no keys.
+     */
+    public List<Jwk> getKeys() {
+        return keys;
+    }
+
+    /**
+     * Get the key in this JWK set matching the key ID.
+     * @param keyId The key ID (kid) of the JWK to get.
+     * @return The JWK or {@code null} if no JWK is found.
+     */
+    public Jwk getKey(String keyId) {
+        for (Jwk key : keys) {
+            if (key.getId() != null && key.getId().equals(keyId)) {
+                return key;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get the key in this JWK set matching the key ID. If the key ID is null and there is only one JWK in this set,
+     * that JWK will be returned.
+     *
+     * <p>This method enables backwards compatibility that allows the only JWK to be returned if the specified keyId is
+     * {@code null}</p>
+     *
+     * @param keyId The key ID (kid) of the JWK to get. If {@code null} and the set contains one Jwk, that JWK will be returned.
+     * @return The JWK or {@code null} if the set contains no JWKs.
+     */
+    Jwk getKeyByIdOrSingleEntry(String keyId) {
```

Comment:
Why isn't this part of the `getKey` method? (it was like that before). Who will be calling this, given that it's pkg-priv, and in what scenario? 
I didn't get the backwards compat comment: This is a whole new class, and this legacy logic is being put in a new method. Devs wanting to keep this usage (I don't see why we would like to keep it tbh) would need to read the migration guide and make changes to access it.

Reply:
We can move that logic to the `getKey` method. If we want to change the behavior going forward (which would be a breaking change), we can do that in a separate PR. The backwards-compatible logic will allow a key to be found from the JwkSet in a non-breaking way.

Reply:
If we need to remove this legacy logic let's do that now while we're on major 0. We should enforce obtaining a key given the kid, and as a fallback, expose another method to return the whole list of keys for the dev to be responsible for finding the appropriate one.

Reply:
That makes sense. When we expose the `JwkSet` on the `JwkProvider` that will enable developers to get the whole list of keys (currently there is `getAll` on the `UrlJwkProvider`, but that is not exposed on the interface). 

I think it makes sense to change the legacy logic in a separate PR, then it's isolated and can be properly labeled. WDYT?

Reply:
On second thought, the use of `kid` is optional. We actually [changed this library](https://github.com/auth0/jwks-rsa-java/issues/23) to stick to that a while ago, so removing that now might look odd. What are your thoughts? I suggested making `getKey(kid)` expect non-null `kid` and then a `getAll()` method for those who need to obtain a key but don't have a `kid`. Auth0's use case will always set a `kid`.


And if you need to remove something, please do that on separate PRs, for changelog labeling and better communication 👍